### ISSUE TITLE
Add optimization to avoid copying a struct if passed by reference and there are no writes to and no reads after passed to a callee.

### ISSUE THREAD

**Issue by LLITCHEV** [dotnet/coreclr#1636]
On Windows structs are allocated on the caller stack and passed by ref to the callee.

If a struct is passed to a callee (callee1), which passes it to another callee (calee2), and there are no writes before the call to callee2, and there are no reads after passing to the callee2, the callee1 could pass to calee2 the reference that was passed to it, without need to allocate a new struct on the callee1 stack and copy the data.

**Comment by omariom** [comment]
Is it similar to https://github.com/dotnet/coreclr/issues/1133?

**Comment by LLITCHEV** [comment]
@omariom Yes... It is very similar. This optimization should result in better codegen for handful of cases not described in 1133, I think.

Thank you for pointing out the existing Issue and linking them together.

**Comment by CarolEidt** [comment]
This is really a distinct issue from #1133. In that issue, what's at issue is that:
- First, the inliner creates copies for all struct parameters, and
- Second, value numbering is not fully enabled for structs, so the copies do not get eliminated.

Addressing this issue requires a different kind of analysis, to determine that the struct being passed is not modified, and is not read after the call to callee1.

A repro, in particular one that reflects an actual real-world scenario, would be useful in prioritizing this.
